fmt/mod.c

```diff
@@ -396,3 +467,192 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
 {
 	return fmt_mod_load_song(song, fp, lflags, 1);
 }
+
+/* incomplete 31 sample M.K. amiga mod saving routine */
+int fmt_mod_save_song(disko_t *fp, song_t *song)
+{
+	uint8_t mod_songtitle[20];
+	uint8_t mod_sampleheader[30];
+	uint8_t mod_orders[128];
+	uint8_t tmp[128];
+	uint8_t mod_pattern[1024];
+
+	int nord, nsmp, maxpat, jmax, joutpos;
+	int i, j, n, period;
+	unsigned int warn = 0;
+	song_note_t *m;
+
+	if (song->flags & SONG_INSTRUMENTMODE)
+		warn |= 1 << WARN_INSTRUMENTS;
+	if (song->flags & SONG_LINEARSLIDES)
+		warn |= 1 << WARN_LINEARSLIDES;
+
+	nsmp = csf_get_num_samples(song); // Getting number of samples
+	if (nsmp > 31) {
+		nsmp = 31;
+		warn |= 1 << WARN_MAXSAMPLES;
+	}
+
+	if (3 < csf_get_highest_used_channel(song))
+		warn |= 1 << WARN_MAXCHANNELS;
+
+	memcpy(mod_songtitle, song->title, 20);
+	disko_write(fp, mod_songtitle, 20); // writing song title
+
+	// Now writing sample headers
+	for(n = 1; n <= 31; ++n) {
+		for(i = 0; i < 30; ++i) mod_sampleheader[i] = 0;
+		if(n <= nsmp) {
+			if(song->samples[n].global_volume != 64)
+				warn |= 1 << WARN_SAMPLEVOL;
+			if((song->samples[n].flags & (CHN_LOOP | CHN_PINGPONGLOOP)) == (CHN_LOOP | CHN_PINGPONGLOOP) || (song->samples[n].flags & CHN_SUSTAINLOOP))
+				warn |= 1 << WARN_LOOPS;
+			if(song->samples[n].vib_depth != 0)
+				warn |= 1 << WARN_SAMPLEVIB;
+			memcpy(mod_sampleheader, song->samples[n].name, 22); // sample name
+			mod_sampleheader[22] = song->samples[n].length >> 9; // sample 11th word MSB length/2
+			mod_sampleheader[23] = song->samples[n].length >> 1; // sample 11th word LSB length/2
+			for(j = 15; j && (finetune_table[j] > song->samples[n].c5speed); --j)
+				if(((song->samples[n].c5speed) > 10000) && (j == 8))
+					break; // determine from finetune_table entry
+			mod_sampleheader[24] = (j ^ 8) & 0x0f; // sample 24th byte finetune value
+			mod_sampleheader[25] = (song->samples[n].volume + 1) / 4; // sample 25th byte sample volume value 0..64 scaled
+			if( song->samples[n].flags & CHN_LOOP ) {
+				mod_sampleheader[26] = song->samples[n].loop_start >> 9;// loop start MSB /2
+				mod_sampleheader[27] = song->samples[n].loop_start >> 1;// loop start LSB /2
+				mod_sampleheader[28] = (song->samples[n].loop_end - song->samples[n].loop_start) >> 9;// loop length MSB /2
+				mod_sampleheader[29] = (song->samples[n].loop_end - song->samples[n].loop_start) >> 1;// loop length LSB /2
+			}
+		}
+		disko_write(fp, mod_sampleheader, 30); // writing current sample header
+	}
+
+	tmp[0] = nord = csf_get_num_orders(song); // or "csf_get_num_orders(song_t *csf);"
+	tmp[1] = 0x7f;
+	disko_write(fp, tmp, 2);
+
+	for(maxpat = i = 0; (i < nord) && (i < 128); ++i) {
+		mod_orders[i] = song->orderlist[i];
+		if(maxpat < mod_orders[i]) maxpat = mod_orders[i];
+	}
+	for(; i < 128; ++i)
+		mod_orders[i] = 0;
+
+	disko_write(fp, mod_orders, 128);
+
+	disko_write(fp, valid_tags[0][0], 4);
```

Comment:
This should write M!K! if the highest pattern index is > 64.

Reply:
done applying >= 64 as barrier